File: entity.py
import math

from typing import Generator, List, Optional, Tuple


class Entity:
    def __init__(
        self,
        cells: Optional[List[Tuple[int, int]]] = None,
        entities: Optional[List["Entity"]] = None,
    ):
        if cells is None and entities is None:
            raise ValueError("Either cells or entities must be passed")

        self._cells = cells
        if cells is None:
            self._cells = []

        if entities is not None:
            for entity in entities:
                for cell in entity.iterator():
                    self._cells.append(cell)

    def iterator(self) -> Generator[Tuple[int, int], None, None]:
        for cell in self._cells:
            yield cell

    def transform(
        self,
        row: int = 0,
        col: int = 0,
        flip: bool = False,
        rotate: int = 0,
    ) -> "Entity":
        """
        1. Flip horizontally
        2. Rotate about the origin clockwise by 90˚
        3. Translate with rows and cols
        """
        cells = []
        for r, c in self._cells:
            # Flip
            if flip:
                c = -c

            # Rotate
            theta = rotate * math.pi / 2
            cos = int(math.cos(theta))
            sin = int(math.sin(theta))
            c, r = cos * c - sin * r, cos * r + sin * c

            # Translate
            r += row
            c += col

            cells.append((r, c))
        return Entity(cells)

File: scripts/entity.ts
import Location from "./location.js";

class Entity {
  name: string;
  locations: Location[];

  constructor(name: string, locations: Location[], entities: Entity[]) {
    this.name = name;
    this.locations = locations;

    entities.forEach((entity) => {
      entity.forEach((location) => {
        this.locations.push(location);
      });
    });
  }

  forEach(callback: (location: Location) => void) {
    this.locations.forEach(callback);
  }

  transform(row: number = 0, col: number = 0, flip: boolean = false, rotate: number = 0): Entity {
    // 1. Flip horizontally
    // 2. Rotate about the origin clockwise by 90˚
    // 3. Translate with rows and cols

    const locations = [];
    this.locations.forEach((location) => {
      let [r, c] = location;

      // Flip
      if (flip) {
        c = -c;
      }

      // Rotate
      const theta = (rotate * Math.PI) / 2;
      const cos = Math.floor(Math.cos(theta));
      const sin = Math.floor(Math.sin(theta));
      [c, r] = [cos * c - sin * r, cos * r + sin * c];

      // Translate
      r += row;
      c += col;

      locations.push([r, c]);
    });
    return new Entity(this.name, locations, []);
  }
}

export default Entity;
